### PR TITLE
nix: windows build

### PR DESCRIPTION
**Pull request by hutli**
Fixes correct file extensien when using pkgsCross stdenv for crossCompiling for windows using mingwW64 inside Nix.

### CODE REVIEW

File: .devops/nix/package.nix [hunk 1]
```diff
@@ -198,6 +201,9 @@ effectiveStdenv.mkDerivation (
         (cmakeBool "LLAMA_VULKAN" useVulkan)
         (cmakeBool "LLAMA_STATIC" enableStatic)
       ]
+      ++ optionals crossCompileWindows [
+        (cmakeFeature "CMAKE_SYSTEM_NAME" "Windows")
+      ]
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:206]: Is https://github.com/NixOS/nixpkgs/blob/13a8da967119594974797df76e0fbf43194303b1/pkgs/build-support/lib/cmake.nix#L8-L9 not enough?

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:1]: Awesome, thanks!

File: .devops/nix/package.nix [hunk 2]
```diff
@@ -222,8 +228,8 @@ effectiveStdenv.mkDerivation (
     # TODO(SomeoneSerge): It's better to add proper install targets at the CMake level,
     # if they haven't been added yet.
     postInstall = ''
-      mv $out/bin/main $out/bin/llama
-      mv $out/bin/server $out/bin/llama-server
+      mv $out/bin/main${executableSuffix} $out/bin/llama
+      mv $out/bin/server${executableSuffix} $out/bin/llama-server
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:232]: This has been brought up before, it's really awkward we maintain a custom intsallPhase: the cmakelists need to be updated

File: .devops/nix/package.nix (hunk 1 above)
Comment (hutli) [reply .devops/nix/package.nix:206]: Oh yes, you are completely correct; this is not needed anymore after I've just started overwriting stdenv. Well spottet, thank you!

File: build.zig [hunk 3]
```diff
@@ -105,20 +105,6 @@ const Maker = struct {
         e.want_lto = m.enable_lto;
         return e;
     }
-
-    fn lib(m: *const Maker, name: []const u8, src: []const u8, deps: []const *Compile) *Compile {
```
Comment (SomeoneSerge) [inline build.zig]: I'm not familiar with llama.cpp's zig codebase, what is the motivation for these changes?

Comment (SomeoneSerge) [reply build.zig]: Oh I guess they're reverted in the latter commits. Maybe rebase `--autosquash`

File: .devops/nix/package.nix [hunk 4]
```diff
@@ -65,6 +65,9 @@ let
     strings.optionalString (suffices != [ ])
       ", accelerated with ${strings.concatStringsSep ", " suffices}";
 
+  crossCompileWindows = (effectiveStdenv.buildPlatform.isLinux || effectiveStdenv.buildPlatform.isDarwin) && effectiveStdenv.targetPlatform.isMinGW;
+  executableSuffix = strings.optionalString (crossCompileWindows) ".exe";
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:69]: Isn't it enough to test just `isMinGW` now? Also it's actually enough to only test the `hostPlatform` I believe

File: .devops/nix/package.nix [hunk 5]
```diff
@@ -222,8 +225,8 @@ effectiveStdenv.mkDerivation (
     # TODO(SomeoneSerge): It's better to add proper install targets at the CMake level,
     # if they haven't been added yet.
     postInstall = ''
-      mv $out/bin/main $out/bin/llama
-      mv $out/bin/server $out/bin/llama-server
+      mv $out/bin/main${executableSuffix} $out/bin/llama
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:228]: Do we really want to remove the `.exe` when targeting windows?

File: build.zig (hunk 3 above)
Comment (hutli) [reply build.zig]: Sorry for the extremely confusing commit history - I went through quite a few experiments before figuring out that we can essentially just replace stdenv gcc compiler with MinGw when cross compiling for Windows. Initially I tried using the Zig build system to achieve this but using the CMake build system is much smarter.

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:1]:
```
❯ nom build github:ggerganov/llama.cpp/553aafca25ed6aa3721b3e0bfa121eed3f2b7b58#windows
❯ wine ./result/bin/llama-bench.exe
007c:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
007c:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
007c:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
007c:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
Application could not be started, or no application associated with the specified file.
ShellExecuteEx failed: Bad EXE format for Z:\home\ss\result\bin\llama-bench.exe
```

Do you get the same errors? I haven't got any windows VMs running rn so I could only test wine

File: .devops/nix/package.nix (hunk 4 above)
Comment (hutli) [reply .devops/nix/package.nix:69]: Ah yes, as we only use this variable for checking which extension to use it is enough to just check our build target. I don't quite understand why you would use `hostPlatform` instead of `buildTarget` though?

File: .devops/nix/package.nix
Comment (SomeoneSerge) [reply .devops/nix/package.nix:1]: Oh! I used `nixpkgs#wine` instead of `nixpkgs#wine64`

Comment (SomeoneSerge) [inline .devops/nix/package.nix:1]:
```
❯ nix run nixpkgs#wine64 -- "$(nix build github:ggerganov/llama.cpp/553aafca25ed6aa3721b3e0bfa121eed3f2b7b58#windows --print-out-paths)"/bin/llama-bench.exe
0084:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
0084:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
0084:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
0084:fixme:hid:handle_IRP_MN_QUERY_ID Unhandled type 00000005
| model                          |       size |     params | backend    |    threads | test       |
         t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ----
-----------: |
main: error: failed to load model 'models/7B/ggml-model-q4_0.gguf'
```


I guess this works

File: .devops/nix/package.nix (hunk 5 above)
Comment (hutli) [reply .devops/nix/package.nix:228]: Valid point - we would probably like to have the `executableSuffix` on the right side of the copy too.

File: build.zig (hunk 3 above)
Comment (hutli) [reply build.zig]: You are more than welcome to just squash all the commits 👍

File: .devops/nix/package.nix (hunk 4 above)
Comment (SomeoneSerge) [reply .devops/nix/package.nix]: H'm? `hostPlatform` is the one on which we'll be running the resulting `llama.exe`. Coincidentally, `targetPlatform` will be the same in all meaningful instantiations of nixpkgs, but I think `hostPlatform.isMinGW` is the one we want?

File: build.zig (hunk 3 above)
Comment (SomeoneSerge) [reply build.zig]: Fair

File: .devops/nix/package.nix (hunk 4 above)
Comment (hutli) [reply .devops/nix/package.nix]: Comitted

**Review by SomeoneSerge — APPROVED** [review]
Let's wait for the CI to pass and **squash**-merge this

File: flake.nix [hunk 6]
```diff
@@ -155,6 +155,9 @@
               {
                 default = config.legacyPackages.llamaPackages.llama-cpp;
                 vulkan = config.packages.default.override { useVulkan = true; };
+                windows = config.legacyPackages.llamaPackages.llama-cpp.override {
+                  stdenv = pkgs.pkgsCross.mingwW64.stdenv;
+                };
```
Comment (SomeoneSerge) [inline flake.nix:160]: Oh wait! Actually, I'm surprised this worked. I never touched windows, but shouldn't we be taking the other dependencies from `pkgsCross` as well?

Comment (SomeoneSerge) [reply flake.nix:160]: Which would look something like `(pkgsCross.mingwW64.callPackage .devops/nix/scope.nix { }).llama-cpp`

Comment (hutli) [reply flake.nix:160]: From my experimentation the only parts that are needed to cross compile for Windows is the MinGW executable and the libraries it comes with which is all in stdenv. I did consider using `pkgsCross.callPackage` instead of overwriting `stdenv` but wouldn't that mean that we cannot compile for windows _and_ CUDA/ROCM since we use `pkgsCuda.callPackage` and `pkgsRocm.callPackage` respectively?

Comment (SomeoneSerge) [reply flake.nix:160]: > but wouldn't that mean that we cannot compile for windows and CUDA/ROCM since we use pkgsCuda.callPackage and pkgsRocm.callPackage respectively?

Being able to construct a configuration with cuda+rocm and exposing a flake output for that configuration are two different things. The flake outputs are there mostly for show. The actual "useful" program is the `package.nix` (and the `scope.nix`) which are functions that you can call from arbitrary instances of Nixpkgs.

Those can be consumed e.g. as overlays

Comment (SomeoneSerge) [reply flake.nix:160]: But also linking the build platform's (linux's) CUDA libraries to a mingw executable would obviously fail

Comment (SomeoneSerge) [reply flake.nix:160]:
```
inputs.nixpkgs = ...
inputs.llama-cpp = ...
outputs = { nixpkgs, llama-cpp, ... }: {
  packages.x86_64-linux.llamaWindowsCuda = (import nixpkgs { config.cudaSupport = true; overlays = [ llama-cpp.overlays.default ]; }).pkgsCross.llama-cpp;
}
```

et cetera (almost certainly broken, we haven't looked into the windows instances of cudaPackages yet)

Comment (hutli) [reply flake.nix:160]: Ah, ok. So using `pkgsCross.mingwW64.callPackage` would still work? I'll look into doing that then.

Comment (SomeoneSerge) [reply flake.nix:160]: You can extend `legacyPackages` the same way it's done for cuda/rocm

Comment (hutli) [reply flake.nix:160]: Yes I've tried doing this while using `pkgs.pkgsCross.mingwW64.callPackage` but I'm getting problems with packages that complain about the platform now. This will take some time to get working in the right way but I'm on it.

Comment (SomeoneSerge) [reply flake.nix:160]: (I'm also surprised we didn't have to use `makeScopeWithSplicing`)

File: .devops/nix/package.nix [hunk 7]
```diff
@@ -65,6 +65,8 @@ let
     strings.optionalString (suffices != [ ])
       ", accelerated with ${strings.concatStringsSep ", " suffices}";
 
+  executableSuffix = strings.optionalString (effectiveStdenv.hostPlatform.isMinGW) ".exe";
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:68]:
```
nix-repl> pkgsCross.mingwW64.hostPlatform.extensions.executable
".exe"
```

File: flake.nix (hunk 6 above)
Comment (hutli) [reply flake.nix:160]: Alright after quite a bit of testing it seems that the best solution is to override `stdenv` - I've tested the binaries on a Windows system and they all seems to run correctly 👍

I think this should be ready to merge. Let me know if you think anything is missing.

File: .devops/nix/package.nix (hunk 7 above)
Comment (hutli) [reply .devops/nix/package.nix]: Wau, that is smart - implemented!

File: flake.nix (hunk 6 above)
Comment (SomeoneSerge) [reply flake.nix]: Hi, check out the patches I just pushed. The only thing that was broken in the cross-compilation set up was python

Comment (hutli) [reply flake.nix]: You did it! I tried to use `pkgs.pkgsCross.mingwW64.callPackage` but I kept getting that python error - which you've fixed now. I can even compile for Vulkan and it works on our Windows machine.

screenshot

**Comment by hutli** [comment]
> @hutli I squashed and force-pushed your commits because there were some git conflicts with the master branch. I imagine you still have the commits with your original experiments (zig&c) locally if that matters

Amazing, thank you! Yes, I have it locally and I'll save it just in case, thank you for the heads up :+1:

File: .devops/nix/package.nix [hunk 8]
```diff
@@ -200,7 +200,7 @@ effectiveStdenv.mkDerivation (
       ++ optionals useMpi [ mpi ]
       ++ optionals useOpenCL [ clblast ]
       ++ optionals useRocm rocmBuildInputs
-      ++ optionals useBlas [ blas ]
+      ++ optionals (useBlas && builtins.elem effectiveStdenv.hostPlatform.system blas.meta.platforms) [ blas ]
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:203]: I think this is `lib.meta.availableOn` but first say, does this really break in `.#windows`?

File: .devops/nix/package.nix [hunk 9]
```diff
@@ -200,7 +200,7 @@ effectiveStdenv.mkDerivation (
       ++ optionals useMpi [ mpi ]
       ++ optionals useOpenCL [ clblast ]
       ++ optionals useRocm rocmBuildInputs
-      ++ optionals useBlas [ blas ]
+      ++ optionals (useBlas && blas.meta.available) [ blas ]
```
Comment (SomeoneSerge) [inline .devops/nix/package.nix:203]: Also I think we should instead `&&` this in https://github.com/ggerganov/llama.cpp/blob/ae316601c60a880d255d877f167eae7a58cc2e05/.devops/nix/package.nix#L21-L27.

Rationale: user-supplied `useBlas` should have the priority, but we want to choose a different default for `useBlas` in the cross-compilation setting

**Comment by SomeoneSerge** [comment]
(the failure is from https://github.com/ggerganov/llama.cpp/issues/6346)
